### ISSUE TITLE
sql: DROP DATABASE does not complete

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#40411]
**Describe the problem**

- running v19.1.4
- user issued 3x DROP DATABASE
- initially all 3 job remain with status "Waiting for GC TTL"
- status "Waiting for GC TTL" persists for more than 25 hours (the configured TTL)
- after 2 days, 2 of the 3 jobs complete, but one remains

Output of SHOW JOBS:
```
> show jobs;
job_id | job_type | description | statement | user_name | status | running_status | created | started | finished | modified | fraction_completed | error | coordinator_id
+--------------------+---------------+---------------------------------+-----------+-----------+-----------+--------------------+----------------------------------+----------------------------------+----------------------------------+----------------------------------+--------------------+-------+----------------+
481077470951407617 | SCHEMA CHANGE | DROP DATABASE xxx1 CASCADE | | root | running | waiting for GC TTL | 2019-08-27 05:26:27.229924+00:00 | 2019-08-27 05:26:33.287607+00:00 | NULL | 2019-08-28 10:15:51.706498+00:00 | 0 | | NULL
481077241315426305 | SCHEMA CHANGE | DROP DATABASE xxx2 CASCADE | | root | succeeded | NULL | 2019-08-27 05:25:20.538785+00:00 | 2019-08-27 05:25:23.206072+00:00 | 2019-08-28 07:11:57.472115+00:00 | 2019-08-28 07:11:57.472116+00:00 | 1 | | NULL
481077135248326657 | SCHEMA CHANGE | DROP DATABASE xxx3 CASCADE | | root | succeeded | NULL | 2019-08-27 05:24:48.596486+00:00 | 2019-08-27 05:24:50.836629+00:00 | 2019-08-28 07:11:56.603044+00:00 | 2019-08-28 07:11:56.603045+00:00 | 1 | | NULL
(3 rows)

Time: 60.223985ms
```

Additional context: 
> I was also decommissioning nodes before I dropped these DB's but the decommission was finished and there where no under-replicated ranges.

**Expected behavior**
Jobs complete successfully

**Additional data / screenshots**

- admin UI screen for databases now displays error "internal server error"
- also did a rolling restart of the cluster (which worked fine) to see if that would alleviate any stuckness - problem persisted
- range replication badness:

  > before the restart I was seeing some weirdness in the UI with regards to "under replicated ranges" and "unavailable ranges". It was periodically showing different numbers. The numbers where sometimes around 2000 for under/unavailable and then a refresh of the UI would show 0 again and I know for a fact there is no way the cluster would recover that quick. Also often node 2 would turn yellow but again the process on the server was up and running. I think it is all somehow related but I am not sure.

- there may have been a zone config constraint on database xxs1 (the one that is stuck)

*Debug zip available* from gdrive link inside zendesk issue 3608

**Additional context**

> If I cant drop the database successfully I am just going to recreate the cluster and put this down to experience.

**Comment by knz** [comment]
cc @dt for triage.
Note: this is a P-3 issue from a customer (K).

**Comment by RoachietheSupportRoach** [comment]
Zendesk ticket #[3608](https://cockroachdb.zendesk.com/agent/tickets/3608) has been linked to this issue.

**Comment by lucy-zhang** [comment]
This is possibly related to the problem described in #38617: A schema change has completed, but updating the job state at the end fails. In that case, due to that PR, we continue instead of retrying endlessly when this happens, so the job becomes orphaned and stuck in the `waiting for GC TTL` state indefinitely. There are log messages at around `2019-08-28 07:11` (the same time as when the two successful jobs finished) indicating that the job state update for the stuck job `481077470951407617` failed:
```
W190828 07:11:54.382827 718 sql/schema_changer.go:386  [n20,scExec] failed to update job 481077470951407617: TransactionRetryWithProtoRefreshError: WriteTooOldError: write at timestamp 1566976314.132394656,0 too old; wrote at 1566976314.132394656,2
W190828 07:11:55.026295 1213 sql/schema_changer.go:386  [n19,scExec] failed to update job 481077470951407617: TransactionRetryWithProtoRefreshError: ReadWithinUncertaintyIntervalError: read at time 1566976314.956689812,0 encountered previous write with future timestamp 1566976314.967198138,0 within uncertainty interval `t <= 1566976314.986089878,0`; observed timestamps: [{11 1566976314.973283453,0} {13 1566976314.986089878,0} {19 1566976314.956689812,0}]
```
These errors are coming from https://github.com/cockroachdb/cockroach/blob/release-19.1/pkg/sql/schema_changer.go#L386, after we've already cleared the table ranges, and are about to delete the table descriptor (and zone configs).

It also looks like not all the table descriptors were actually successfully deleted, though, so the above problem doesn't fully account for the job being in this state. According to `system.jobs`, the tables to be dropped in that database were `{109,115,108,107,110,117,118,111,116,113}`. Later on, the lease manager attempts to refresh those same table leases for some reason:
```
I190828 19:20:54.581969 29860418 sql/lease.go:1790  refreshing table: 120 lease failed: descriptor not found
I190828 19:20:54.582060 29860417 sql/lease.go:1790  refreshing table: 122 lease failed: descriptor not found
I190828 19:20:54.588320 29860395 sql/lease.go:1790  refreshing table: 108 lease failed: descriptor not found
I190828 19:20:54.588385 29860177 sql/lease.go:1790  refreshing table: 97 lease failed: descriptor not found
I190828 19:20:54.594604 29860454 sql/lease.go:1790  refreshing table: 109 lease failed: descriptor not found
I190828 19:20:54.618182 29860319 sql/lease.go:1790  refreshing table: 116 lease failed: descriptor not found
I190828 19:20:54.628698 29860487 sql/lease.go:1790  refreshing table: 113 lease failed: descriptor not found
I190828 19:20:54.636413 29860078 sql/lease.go:1790  refreshing table: 107 lease failed: descriptor not found
I190828 19:20:54.636760 29860490 sql/lease.go:1790  refreshing table: 89 lease failed: descriptor not found
I190828 19:20:54.642827 29860340 sql/lease.go:1790  refreshing table: 117 lease failed: table is being dropped
I190828 19:20:54.646421 29860517 sql/lease.go:1790  refreshing table: 118 lease failed: descriptor not found
I190828 19:20:54.652250 29860519 sql/lease.go:1790  refreshing table: 100 lease failed: descriptor not found
I190828 19:20:54.665921 29860403 sql/lease.go:1790  refreshing table: 111 lease failed: descriptor not found
I190828 19:20:54.671933 29860535 sql/lease.go:1790  refreshing table: 115 lease failed: descriptor not found
I190828 19:20:54.672127 29860549 sql/lease.go:1790  refreshing table: 96 lease failed: descriptor not found
I190828 19:20:54.681657 29860492 sql/lease.go:1790  refreshing table: 93 lease failed: descriptor not found
I190828 19:20:54.681681 29860521 sql/lease.go:1790  refreshing table: 110 lease failed: descriptor not found
```
So table 117 apparently hasn't had its descriptor deleted. Looking at some of the range span keys, it seems that table 117 was interleaved into 116, which may be relevant.

Unfortunately debug.zip doesn't contain schemas/descriptors for dropped tables.

**Comment by lucy-zhang** [comment]
Looking at a backup of `system.descriptor`, none of the database or table descriptors for any of the dropped databases (including the one corresponding to the stuck job) are present. This suggests that the problem is that the job is now orphaned because we didn't update its state successfully. I opened #40563 to track this. There's no fix in 19.2, though I think improvements in this area are planned for 20.1.

There seems to be another problem, though, which is that some tables (ranges, descriptors) from the stuck job seem to have remained, at least for some amount of time. The logs I posted above show that table 117's descriptor still existed at that point (in the dropping state), and the range lists from debug.zip have many populated ranges for table 116 (which table 117 is interleaved into), though all the other table spans (based on the range lists and the logs) seem to have been successfully cleared. There may be a problem with dropping interleaved tables.
